File: src/vca/storage.py
"""对话记录存储 — 自动保存/恢复会话"""

from __future__ import annotations

import json
import os
import time
from datetime import datetime, timezone, timedelta
from pathlib import Path

from langchain_core.messages import (
    BaseMessage,
    HumanMessage,
    AIMessage,
    SystemMessage,
    ToolMessage,
)

# ---- 路径常量 ----
_BASE_DIR = Path.home() / ".vca"
_SESSIONS_DIR = _BASE_DIR / "sessions"
_INDEX_FILE = _BASE_DIR / "session_index.json"

# ---- 时区 ----
_TZ = timezone(timedelta(hours=8))  # UTC+8


def _ensure_dirs() -> None:
    _SESSIONS_DIR.mkdir(parents=True, exist_ok=True)


# ============================================================
# 序列化 / 反序列化
# ============================================================

def _message_to_dict(msg: BaseMessage) -> dict:
    """将 LangChain 消息转为可 JSON 序列化的 dict"""
    out: dict = {"type": type(msg).__name__}
    if isinstance(msg, HumanMessage):
        out["content"] = msg.content
    elif isinstance(msg, AIMessage):
        out["content"] = msg.content
        # 仅当有 tool_calls 时才写入，避免 null 导致反序列化校验失败
        if getattr(msg, "tool_calls", None):
            out["tool_calls"] = list(msg.tool_calls)
    elif isinstance(msg, SystemMessage):
        out["content"] = msg.content
    elif isinstance(msg, ToolMessage):
        out["content"] = msg.content
        out["tool_call_id"] = msg.tool_call_id
        out["name"] = getattr(msg, "name", "")
    return out


def _dict_to_message(data: dict) -> BaseMessage:
    """从 dict 还原为 LangChain 消息"""
    msg_type = data.get("type", "")
    content = data.get("content", "")
    if msg_type == "HumanMessage":
        return HumanMessage(content=content)
    elif msg_type == "AIMessage":
        tool_calls = data.get("tool_calls")
        # tool_calls 为 None 时不要传参，否则新版 langchain 校验失败
        if tool_calls:
            return AIMessage(content=content, tool_calls=tool_calls)
        return AIMessage(content=content)
    elif msg_type == "SystemMessage":
        return SystemMessage(content=content)
    elif msg_type == "ToolMessage":
        return ToolMessage(
            content=content,
            tool_call_id=data.get("tool_call_id", ""),
            name=data.get("name", ""),
        )
    # fallback
    return HumanMessage(content=str(data))


def _serialize_messages(messages: list[BaseMessage]) -> list[dict]:
    """序列化消息列表"""
    return [_message_to_dict(m) for m in messages if not _is_internal_message(m)]


def _deserialize_messages(data: list[dict]) -> list[BaseMessage]:
    """反序列化消息列表"""
    return [_dict_to_message(d) for d in data]


def _is_internal_message(msg: BaseMessage) -> bool:
    """过滤内部消息 (ask_user 占位等)"""
    if isinstance(msg, ToolMessage):
        if msg.content == "[AWAITING_USER_INPUT]":
            return True
    return False


# ============================================================
# 索引管理
# ============================================================

def _read_index() -> list[dict]:
    """读取会话索引"""
    _ensure_dirs()
    try:
        if _INDEX_FILE.exists():
            return json.loads(_INDEX_FILE.read_text(encoding="utf-8"))
    except Exception:
        pass
    return []


def _write_index(index: list[dict]) -> None:
    """写入会话索引"""
    _ensure_dirs()
    _INDEX_FILE.write_text(
        json.dumps(index, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )


def _update_index(
    session_id: str, title: str, workspace: str, message_count: int
) -> None:
    """更新索引中的一条记录 (不存在则新增)"""
    now = datetime.now(_TZ).strftime("%Y-%m-%d %H:%M")
    entry = {
        "id": session_id,
        "title": title,
        "workspace": workspace,
        "message_count": message_count,
        "updated_at": now,
    }

    index = _read_index()
    for i, item in enumerate(index):
        if item.get("id") == session_id:
            entry["created_at"] = item.get("created_at", now)
            entry["title"] = item.get("title") or title  # 保留已有标题
            index[i] = entry
            _write_index(index)
            return

    entry["created_at"] = now
    index.insert(0, entry)
    # 只保留最近 50 条
    _write_index(index[:50])


def _remove_from_index(session_id: str) -> None:
    """从索引中删除一条记录"""
    index = _read_index()
    index = [item for item in index if item.get("id") != session_id]
    _write_index(index)


# ============================================================
# 保存 / 加载 / 管理
# ============================================================

def save_session(
    session_id: str,
    messages: list[BaseMessage],
    workspace_dir: str,
    title: str = "",
) -> str:
    """
    保存会话到磁盘。

    Args:
        session_id: 会话 ID
        messages: 消息历史
        workspace_dir: 工作空间路径
        title: 会话标题 (留空则从第一条用户消息提取)

    Returns:
        session_id
    """
    _ensure_dirs()

    # 过滤掉内部消息
    clean = [m for m in messages if not _is_internal_message(m)]

    if not title:
        # 从第一条 HumanMessage 提取标题
        for m in clean:
            if isinstance(m, HumanMessage) and m.content:
                title = m.content[:60].replace("\n", " ")
                break
        title = title or "未命名会话"

    # 序列化
    data = _serialize_messages(clean)

    filepath = _SESSIONS_DIR / f"{session_id}.json"
    filepath.write_text(
        json.dumps(data, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )

    _update_index(session_id, title, workspace_dir, len(clean))

    return session_id


def load_session(session_id: str) -> dict | None:
    """
    加载指定会话。

    Returns:
        {"messages": [...], "workspace_dir": "...", "title": "..."}
        或 None (会话不存在)
    """
    _ensure_dirs()
    filepath = _SESSIONS_DIR / f"{session_id}.json"
    if not filepath.exists():
        return None

    try:
        data = json.loads(filepath.read_text(encoding="utf-8"))
        messages = _deserialize_messages(data)

        # 从索引中查找元信息
        workspace = ""
        title = ""
        index = _read_index()
        for item in index:
            if item.get("id") == session_id:
                workspace = item.get("workspace", "")
                title = item.get("title", "")
                break

        return {
            "messages": messages,
            "workspace_dir": workspace,
            "title": title,
        }
    except Exception as e:
        return None


def list_sessions(max_count: int = 20) -> list[dict]:
    """列出最近的会话记录"""
    index = _read_index()
    return index[:max_count]


def delete_session(session_id: str) -> bool:
    """删除指定会话 (文件 + 索引)"""
    _ensure_dirs()
    filepath = _SESSIONS_DIR / f"{session_id}.json"
    deleted = False
    if filepath.exists():
        filepath.unlink()
        deleted = True
    _remove_from_index(session_id)
    return deleted


def generate_session_id() -> str:
    """生成新的 session ID (时间戳)"""
    return datetime.now(_TZ).strftime("%Y%m%d_%H%M%S_") + f"{int(time.time() * 1000) % 100000:05d}"


def get_last_session() -> dict | None:
    """获取最近的会话"""
    sessions = list_sessions(1)
    if not sessions:
        return None
    return load_session(sessions[0]["id"])


def auto_save(
    state,
    session_id: str | None = None,
) -> str:
    """
    自动保存当前状态。

    Args:
        state: AgentState
        session_id: 已有会话 ID，为 None 则新建

    Returns:
        使用的 session_id
    """
    if session_id is None:
        session_id = generate_session_id()

    save_session(
        session_id=session_id,
        messages=state.get("messages", []),
        workspace_dir=state.get("workspace_dir", ""),
    )
    return session_id

File: src/storage.ts
/**
 * 对话记录存储 —— 自动保存 / 恢复会话。
 *
 * 这一层只干两件事：**把消息与 BaseMessage 互转**（LangChain 的对象图不能直接
 * JSON 化），以及**把路径定下来**（`VCA_DIR` 来自 config.ts）。索引、列表、
 * 搜索、删除那些纯文件逻辑全在 `./session-store.js`，那边目录由调用方传进去，
 * 所以能被单独测（本文件的 `import "./config.js"` 会在 import 时写
 * `~/.vca/config.json`，一行都测不得）。
 *
 * 拆分的边界就是「要不要碰用户的真实家目录」，与 `paths.ts` / `tools/executable.ts`
 * 同一个理由。
 */
import {
  AIMessage,
  HumanMessage,
  SystemMessage,
  ToolMessage,
  type BaseMessage,
} from "@langchain/core/messages";

import { VCA_DIR } from "./config.js";
import type { AgentState } from "./agent/state.js";
import {
  deleteSession as deleteSessionIn,
  generateSessionId as generateSessionIdIn,
  listDefault,
  listSessions as listSessionsIn,
  nowStr,
  readSessionFile,
  searchSessions as searchSessionsIn,
  sessionsDir,
  upsertIndex,
  writeSessionFile,
  type DeleteOutcome,
  type SessionRow,
  type SessionSelection,
} from "./session-store.js";

// ============================================================
// 序列化 / 反序列化
// ============================================================

type MessageDict = {
  type: string;
  /** content 可以是字符串 (纯文本) 或字符串数组 (多模态 parts) */
  content: string | Array<{ type: string; text?: string; image_url?: { url: string } }>;
  tool_calls?: Array<{ name: string; args: Record<string, unknown>; id: string }>;
  tool_call_id?: string;
  name?: string;
};

export function messageToDict(msg: BaseMessage): MessageDict {
  const raw = msg.content;
  let content: MessageDict["content"];
  if (Array.isArray(raw)) {
    // 多模态 content (如含图片的 HumanMessage)
    content = raw as MessageDict["content"];
  } else {
    content = String(raw ?? "");
  }
  const out: MessageDict = { type: msg.constructor.name, content };
  if (msg instanceof AIMessage) {
    if (msg.tool_calls?.length) {
      out.tool_calls = msg.tool_calls.map((tc) => ({
        name: tc.name,
        args: (tc.args ?? {}) as Record<string, unknown>,
        id: tc.id ?? "",
      }));
    }
  } else if (msg instanceof ToolMessage) {
    out.tool_call_id = msg.tool_call_id;
    out.name = msg.name ?? "";
  }
  return out;
}

export function dictToMessage(data: MessageDict): BaseMessage {
  const content = data.content ?? "";
  const type = data.type.replace("Message", "").toLowerCase();
  switch (type) {
    case "human":
      return new HumanMessage({ content: content as never });
    case "ai": {
      const toolCalls = data.tool_calls;
      if (toolCalls?.length) {
        return new AIMessage({
          content,
          tool_calls: toolCalls.map((tc) => ({
            name: tc.name,
            args: tc.args ?? {},
            id: tc.id ?? `call_${Math.random().toString(36).slice(2, 10)}`,
            type: "tool_call",
          })),
        });
      }
      return new AIMessage({ content });
    }
    case "system":
      return new SystemMessage({ content });
    case "tool":
      return new ToolMessage({
        content,
        tool_call_id: data.tool_call_id ?? "",
        name: data.name ?? "",
      });
    default:
      return new HumanMessage({ content: String(data) });
  }
}

function isInternalMessage(msg: BaseMessage): boolean {
  return msg instanceof ToolMessage && msg.content === "[AWAITING_USER_INPUT]";
}

function serializeMessages(messages: BaseMessage[]): MessageDict[] {
  return messages.filter((m) => !isInternalMessage(m)).map(messageToDict);
}

function deserializeMessages(data: MessageDict[]): BaseMessage[] {
  return data.map(dictToMessage);
}

// ============================================================
// 对外 API（路径绑定在 VCA_DIR 上）
// ============================================================

/** 会话文件所在目录（`/history` 的提示里会报出来，用户好自己去备份） */
export function getSessionsDir(): string {
  return sessionsDir(VCA_DIR);
}

/** `/history` 与 `/load` 共用的列表窗口大小 —— 只有一个来源，别在命令里写死数字 */
export function getListWindow(): number {
  return listDefault();
}

export function saveSession(
  sessionId: string,
  messages: BaseMessage[],
  workspaceDir: string,
  title = ""
): string {
  const clean = messages.filter((m) => !isInternalMessage(m));

  if (!title) {
    for (const m of clean) {
      if (m instanceof HumanMessage && m.content) {
        title = String(m.content).replace(/\n/g, " ").slice(0, 60);
        break;
      }
    }
    title = title || "未命名会话";
  }

  const data = serializeMessages(clean);
  writeSessionFile(VCA_DIR, sessionId, data);

  // 时间戳的写法只有一处（session-store.nowStr）—— 这里原来又拼了一遍
  // `new Date(Date.now() + 8*3600*1000).toISOString()...`：索引里的 created_at
  // 与 updated_at 一旦分头改格式，界面上那两个时间就会长得不一样
  const now = nowStr();
  upsertIndex(VCA_DIR, {
    id: sessionId,
    title,
    workspace: workspaceDir,
    message_count: clean.length,
    created_at: now,
    updated_at: now,
  });
  return sessionId;
}

export function loadSession(
  sessionId: string
): { messages: BaseMessage[]; workspace_dir: string; title: string } | null {
  if (!/^[A-Za-z0-9_-]+$/.test(String(sessionId ?? ""))) return null;
  try {
    const data = readSessionFile(VCA_DIR, sessionId) as MessageDict[];
    const messages = deserializeMessages(data);

    let workspace = "";
    let title = "";
    for (const item of listSessionsIn(VCA_DIR)) {
      if (item.id === sessionId) {
        workspace = item.workspace;
        title = item.title;
        break;
      }
    }
    return { messages, workspace_dir: workspace, title };
  } catch {
    return null;
  }
}

/**
 * 会话列表：索引里有的 + **索引外的会话文件**。
 *
 * 不传 `maxCount` 是全部 —— 数总数的地方（窗口号、序号范围）不能被截断，
 * 要截断只在显示时截一次。
 */
export function listSessions(maxCount?: number): SessionRow[] {
  return listSessionsIn(VCA_DIR, maxCount);
}

export function searchSessions(opts: { limit?: number; filter?: string } = {}): SessionSelection {
  return searchSessionsIn(VCA_DIR, opts);
}

export function deleteSession(sessionId: string): DeleteOutcome {
  return deleteSessionIn(VCA_DIR, sessionId);
}

export function generateSessionId(): string {
  return generateSessionIdIn();
}

export function getLastSession(): { messages: BaseMessage[]; workspace_dir: string; title: string } | null {
  const sessions = listSessions(1);
  if (sessions.length === 0) return null;
  return loadSession(sessions[0].id);
}

export function autoSave(state: AgentState, sessionId?: string): string {
  if (!sessionId) sessionId = generateSessionId();
  saveSession(sessionId, state.messages, state.workspace_dir);
  return sessionId;
}
